### PR TITLE
Refactor step class allocation

### PR DESCRIPTION
**Pull request by sb2nov**
- Step class allocation refactor
- Yaml indentation
- Logging level fixed
- Connections refactor
- Log QA steps and pipeline instances

### CODE REVIEW

File: dataduct/etl/utils.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+        # Load source from the file path provided
+        step_mod = imp.load_source(step_type, path)
+
+        # Get the step class based on class_name provided
+        step_class = getattr(step_mod, step_def['class_name'])
+
+        # Check if step_class is of type ETLStep
+        if not issubclass(step_class, ETLStep):
+            raise ETLInputError('Step type %s is not of type ETLStep')
```
Comment (tpcstld) [inline dataduct/etl/utils.py]: What's the %s being filled by?

File: dataduct/data_access/connection.py [hunk 2]
```diff
@@ -7,50 +7,73 @@
...
+    """Get SQL config from config file and return the dictionary
     """
     if not hasattr(config, 'mysql'):
         raise ETLConfigError('mysql not found in dataduct configs')
 
-    if host_name not in config.mysql:
-        raise ETLConfigError('Config for hostname: %s not found' %host_name)
+    if database_name not in config.mysql:
+        raise ETLConfigError(
+            'Config for hostname: %s not found' %database_name)
```
Comment (tpcstld) [inline dataduct/data_access/connection.py]: Should put space after %

File: dataduct/etl/utils.py (hunk 1 above)
Comment (sb2nov) [inline dataduct/etl/utils.py]: good catch, fixing now.

File: dataduct/config/logger_config.py [hunk 3]
```diff
@@ -22,7 +22,10 @@ def logger_configuration():
         log_directory = config.logging.get(
             'LOG_DIR', os.path.join(os.path.expanduser(CONFIG_DIR)))
         file_name = config.logging.get('LOG_FILE', LOG_FILE)
-        console_level = config.logging.get('DEBUG_LEVEL', logging.WARNING)
+        console_level = config.logging.get('CONSOLE_DEBUG_LEVEL',
```
Comment (tpcstld) [inline dataduct/config/logger_config.py]: Should format the four statements like
```
VARIABLE_NAME = config.logging.get(
    'CONFIG_NAME', DEFAULT_VALUE)
```
so it looks consistent and more readable

File: dataduct/config/config.py [hunk 4]
```diff
@@ -75,7 +79,7 @@ def __init__(self, mode=None):
     def __str__(self):
         """String output for the config object
         """
-        return yaml.dump(self._root_config, default_flow_style=False)
+        return yaml.dump(self._root_config, default_flow_style=False, indent=4)
```
Comment (tpcstld) [inline dataduct/config/config.py]: Shouldn't it be `indent=2`?
Our config files are formatted with 2 spaces, see `.travis.yaml`

File: dataduct/data_access/connection.py [hunk 5]
```diff
@@ -7,50 +7,73 @@
...
 
 config = Config()
+CONNECTION_RETRIES = config.etl.get('CONNECTION_RETRIES', 2)
 
-@retry(2, 60)
-def redshift_connection(**kwargs):
-    """Fetch a psql connection object to redshift
+
+def get_redshift_config():
+    """Get redshift config from config file and return the dictionary
```
Comment (tpcstld) [inline dataduct/data_access/connection.py]: We should start strictly enforcing the docstring style, aka Returns and Args

File: dataduct/etl/etl_pipeline.py [hunk 6]
```diff
@@ -86,7 +75,7 @@ def __init__(self, name, frequency='one-time',
             load_hour, load_min = [None, None]
 
         # Input variables
-        self._name = name
+        self._name = name if not NAME_PREFIX else NAME_PREFIX + '_' + name
```
Comment (tpcstld) [inline dataduct/etl/etl_pipeline.py]: Should add a test for this :P

File: dataduct/pipeline/utils.py [hunk 7]
```diff
@@ -169,3 +176,35 @@ def list_pipelines(conn=None):
...
+    """List of instance rows formatted to match
+    """
+    etl_runs = pipeline.instance_details()
+    entries = []
+    for etl_run_dt in sorted(etl_runs.keys()):
+
+        # Look through instances
+        for instance in sorted(
+                etl_runs[etl_run_dt],
+                key=lambda x: x.get('@actualEndTime', None)):
```
Comment (tpcstld) [inline dataduct/pipeline/utils.py]: Use constant names?

File: dataduct/qa/check.py [hunk 8]
```diff
@@ -105,3 +116,42 @@ def publish(self, export_func=None):
                 self.alert_func(self.summary, self.alert_subject)
             else:
                 raise Exception(self.alert_subject)
+
+    def log_output_to_s3(self, destination_sql=None, table=None,
+                         path_suffix=None):
+        """Log the results of the QA test in S3
+        """
+        assert exactly_one(destination_sql, table), "Needs table or dest_sql"
```
Comment (tpcstld) [inline dataduct/qa/check.py]: Should use Exceptions or change lines 114-118 to avoid inconsistency.

File: dataduct/etl/etl_pipeline.py [hunk 9]
```diff
@@ -541,18 +430,17 @@ def create_steps(self, steps_params, is_bootstrap=False):
...
                 raise
 
             try:
                 step = step_class(**step_args)
             except Exception:
-                print "Error creating step of class %s, step_param %s." % (
-                    str(step_class.__name__),
-                    str(step_args)
-                )
+                logger.error("Error creating step of class %s, step_param %s.",
```
Comment (tpcstld) [inline dataduct/etl/etl_pipeline.py]: Single quotes!

File: dataduct/data_access/connection.py (hunk 5 above)
Comment (sb2nov) [inline dataduct/data_access/connection.py]: Yeah I just didn't have enough time, I'll fix them in some other diff going forward.

File: dataduct/config/config.py (hunk 4 above)
Comment (sb2nov) [inline dataduct/config/config.py]: only the travis file is indent=2, all pipeline definitions and other YAMLs in our code base are 4.

**Pull request merged**